oauthenticator/tests/test_oauth2.py

```diff
@@ -206,3 +207,57 @@ async def test_add_user_override(
         assert added_user.name in authenticator.allowed_users
     else:
         assert added_user.name not in authenticator.allowed_users
+
+
+async def test_login_handler_pkce():
+    authenticator = OAuthenticator(pkce=True)
+    login_handler = mock_handler(OAuthLoginHandler, authenticator=authenticator)
+
+    login_handler._generate_state_id = Mock(return_value=TEST_STATE_ID)
+    login_handler.set_state_cookie = Mock()
+    login_handler.authorize_redirect = Mock()
+
+    login_handler.get()  # no await, since authorize_redirect is mocked
+
+    # Check that PKCE parameters are included
+    assert (
+        "code_challenge"
+        in login_handler.authorize_redirect.call_args.kwargs["extra_params"]
+    )
+    assert (
+        login_handler.authorize_redirect.call_args.kwargs["extra_params"][
+            "code_challenge_method"
+        ]
+        == "S256"
+    )
+
+
+async def test_callback_handler_pkce():
```

Comment:
Can you also add a test to verify an error is returned if PKCE is requested but the server doesn't support it?

Reply:
Done. I guess the way an Oauth server tells it doesn't support PKCE would be by returning a 403 when the client tries to exchange the code for a token.

Reply:
Won't it work but just not be checked for validity if the provider doesn't support it (ignored extra parameters)? If that's true, should it be on by default?

Reply:
If the parameters are definitely ignored then we could always send them, and change the property name to `require_pkce` to enforce it on the client side.

Reply:
@renan-r-santos Sorry, I realise now I was unclear in my request. I was thinking of testing that `raise web.HTTPError(400, "Missing code_verifier")` is raised when the server silently ignores the PKCE request.

Regarding whether or not to always send the PKCE request, how about if we rename the parameter `require_pkce` instead of `pkce`, but keep the current implementation (only send the PKCE field when `require_pkce = True`? That lets us switch to always sending PKCE in future if we want, without having to change or add any parameters.

Reply:
> Regarding whether or not to always send the PKCE request, how about if we rename the parameter require_pkce instead of pkce, but keep the current implementation (only send the PKCE field when require_pkce = True?

That sounds like a good plan to me. I've updated the PR to reflect that.

> I was thinking of testing that raise web.HTTPError(400, "Missing code_verifier") is raised when the server silently ignores the PKCE request.

`raise web.HTTPError(400, "Missing code_verifier")` won't get raised if the server silently ignores the PKCE request. `code_verifier` is data that we store in a cookie together with `state_id` and `next_url` in the login handler, but `code_verifier` isn't sent to or returned from the OAuth provider during login. It is only during code exchange that the client grabs the `code_verifier` previously stored in a cookie and sends it to the server so it can hash it and compare it with the `code_challenge`.

So, the error you mentioned can only happen if the cookie got somehow deleted or corrupted between login and callback handlers. If you still think it is worth adding a test for that, let me know and I'll update the PR. I could be wrong, but I don't think there's a way for a client to know if a server ignores PKCE parameters. On the other hand, a server can enforce that clients use PKCE.

Reply:
Yeah, I think PKCE is only for the provider to check; clients only provide information. So the only reason to disable it that I can see is if some provider raises on unrecognized arguments, which is [officially wrong](https://www.rfc-editor.org/rfc/rfc6749#section-3.1):

> The authorization server MUST ignore unrecognized request parameters.

So if we're only talking about valid OAuth providers, we don't even need to make it optional.

From the [PKCE spec](https://datatracker.ietf.org/doc/html/rfc7636#section-5):

> As the OAuth 2.0 [[RFC6749](https://datatracker.ietf.org/doc/html/rfc6749)] server responses are unchanged by this
   specification, client implementations of this specification do not
   need to know if the server has implemented this specification or not
   and SHOULD send the additional parameters as defined in [Section 4](https://datatracker.ietf.org/doc/html/rfc7636#section-4) to
   all servers.

i.e. it's always right to send PKCE, and it's entirely up to the provider to decide whether to validate or not.

Reply:
Thank you very much for the references, especially

> The authorization server MUST ignore unrecognized request parameters.

which I didn't know and makes a lot of difference.